cmake/FindHAL.cmake

```diff
@@ -1,251 +1,84 @@
-    spi sram swpmi tim tsc uart usart wwdg
-)
-set(HAL_EX_DRIVERS_L4
-    adc crc cryp dac dfsdm dma flash hash i2c ltdc mmc opamp pcd pwr rcc rng rtc 
-    sai sd smartcard spi tim uart usart
-)
-set(HAL_LL_DRIVERS_L4
-    adc comp crc crs dac dma dma2d exti fmc gpio i2c lptim lpuart opamp pka pwr 
-    rcc rng rtc sdmmc spi swpmi tim usart usb utils
-)
-
-set(HAL_DRIVERS_L5
-    adc comp cortex crc cryp dac dfsdm dma exti fdcan flash flash_ramfunc gpio 
-    gtzc hash i2c icache irda iwdg lptim mmc nand nor opamp ospi pcd pka pwr rcc 
-    rng rtc sai sd smartcard smbus spi sram tim tsc uart usart wwdg
-)
-set(HAL_EX_DRIVERS_L5
-    adc crc cryp dac dfsdm dma flash hash i2c mmc opamp pcd pwr rcc
-    rng rtc sai sd smartcard spi tim uart usart
-)
-set(HAL_LL_DRIVERS_L5
-    adc comp crc crs dac dma exti fmc gpio i2c lptim lpuart opamp pka pwr rcc 
-    rng rtc sdmmc spi tim ucpd usart usb utils
-)
+# This function gets a list of hal_driver using a given prefix and suffix
+#
+# out_list_hal_drivers   list of hal_drivers foud
+# hal_drivers_path       path to the hal's drivers
+# hal_driver_type        hal_driver type to find (hal/ll/ex)
+
+function(get_list_hal_drivers out_list_hal_drivers hal_drivers_path hal_driver_type)
+    #the file pattern depends on the hal_driver_type field
+    if(${hal_driver_type} STREQUAL "hal" OR ${hal_driver_type} STREQUAL "ll")
+        set(file_pattern ".+_${hal_driver_type}_([a-z0-9]+)\\.c$")
+    elseif(${hal_driver_type} STREQUAL "ex")
+        set(file_pattern ".+_hal_([a-z0-9]+)_ex\\.c$")
+    else()
+        message(FATAL_ERROR "the inputed hal_driver_type(${hal_driver_type}) is not valid.")
+    endif()
 
-set(HAL_DRIVERS_MP1
-    adc cec cortex crc cryp dac dcmi dfsdm dma exti fdcan gpio hash hsem i2c
-    ipcc lptim mdios mdma pwr qspi rcc rng rtc sai sd smartcard smbus spdifrx
-    spi sram tim uart usart wwdg
-)
-set(HAL_EX_DRIVERS_MP1
-    adc crc cryp dac dfsdm dma hash i2c pwr rcc rtc sai sd smartcard spi tim
-    uart usart
-)
-set(HAL_LL_DRIVERS_MP1
-    adc delayblock dma exti fmc gpio i2c lptim pwr rcc rtc sdmmc spi tim usart
-    utils
-)
+    #Retrieving all the .c files from hal_drivers_path
+    file(GLOB filtered_files
+        RELATIVE "${hal_drivers_path}/Src"
+        "${hal_drivers_path}/Src/*.c")
+    #Keeping only matching files
+    list(FILTER filtered_files INCLUDE REGEX ${file_pattern})
+    #Cleaning files name to have the hal_drivers only (adc/uart/...)
+    list(TRANSFORM filtered_files REPLACE ${file_pattern} "\\1")
+    #Outputing resulting list
+    set(${out_list_hal_drivers} ${filtered_files} PARENT_SCOPE)
+endfunction()
+
+#Checking the families provided to find_package(HAL <params>) such as STM32F4
+foreach(COMP ${HAL_FIND_COMPONENTS})
+    string(TOLOWER ${COMP} COMP_L)
+    string(TOUPPER ${COMP} COMP_U)
+    string(REGEX MATCH "^STM32([FGHLMUW]P?[0-9BL])([0-9A-Z][0-9M][A-Z][0-9A-Z])?_?(M0PLUS|M4|M7)?.*$" COMP_U ${COMP_U})
+    if(CMAKE_MATCH_1)
```

Comment:
please add a comment (example) about what is matched. because https://cmake.org/cmake/help/latest/variable/CMAKE_MATCH_n.html